src/transports/tcp.rs

```diff
@@ -526,20 +526,22 @@ fn read_exact_timeout(
 
     let timeout = timeout.unwrap_or(endpoint.config.read_timeout);
 
-    endpoint
-        .stream
-        .set_read_timeout(Some(timeout))
-        .map_err(|e| {
-            PeerNetError::SendError.error("error setting read timeout", Some(e.to_string()))
-        })?;
-
     let mut total_read: usize = 0;
 
     while total_read < data.len() {
-        if start_time.elapsed() >= timeout {
+        let elapsed = start_time.elapsed();
+        let new_timeout = timeout.saturating_sub(elapsed);
+        if elapsed >= timeout || new_timeout.is_zero() {
             return Err(PeerNetError::ReceiveError.error("timeout read data", None));
         }
 
+        endpoint
+            .stream
+            .set_read_timeout(Some(new_timeout))
+            .map_err(|e| {
+                PeerNetError::SendError.error("error setting read timeout", Some(e.to_string()))
```

Comment:
Senderror => ReceiveError

Reply:
it's outdated, i have create CouldNotSetTimeout error